### PR TITLE
use grpc http2 keepalives for storage + compute rpcs

### PR DESCRIPTION
**Pull request by pH14**
In https://github.com/MaterializeInc/materialize/issues/20299 we saw that `envd`'s grpc connections for storage and compute cannot detect a connection failure when there's no TCP reset. This PR adds in knobs for HTTP/2 keep-alives (and a connect timeout) to detect this scenario. @def- already wrote a test for this scenario (thank you!), which this PR now enables / passes.

The actual change is quite small, but wiring up the knobs into `StorageParameters` and `ComputeParameters` so they can be dynamically updated / accessed requires reams of new code. Unfortunately a lot of the boilerplate is to make `GrpcClientParameters` serializable over the network, even though it's only ever needed by `envd`.

### Motivation

Fixes https://github.com/MaterializeInc/materialize/issues/20299 

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/sql/src/session/vars.rs [hunk 1]
```diff
@@ -1102,6 +1102,31 @@ const ENABLE_STORAGE_SHARD_FINALIZATION: ServerVar<bool> = ServerVar {
...
+    pub const HTTP2_KEEP_ALIVE_INTERVAL: ServerVar<Duration> = ServerVar {
+        name: UncasedStr::new("grpc_client_http2_keep_alive_interval"),
+        value: &Duration::from_secs(3),
+        description: "Idle time to wait before sending HTTP/2 PINGs to maintain established gRPC client connections.",
+        internal: true,
+    };
+    pub const HTTP2_KEEP_ALIVE_TIMEOUT: ServerVar<Duration> = ServerVar {
+        name: UncasedStr::new("grpc_client_http2_keep_alive_timeout"),
+        value: &Duration::from_secs(5),
+        description:
```
Comment (guswynn) [inline src/sql/src/session/vars.rs:1124]: curious, is it considered normal for the timeout to be longer than the interval?

File: src/storage-client/src/controller/rehydration.rs [hunk 2]
```diff
@@ -68,6 +69,7 @@ where
         build_info: &'static BuildInfo,
         metrics: RehydratingStorageClientMetrics,
         envd_epoch: NonZeroI64,
+        grpc_client_params: GrpcClientParameters,
```
Comment (guswynn) [inline src/storage-client/src/controller/rehydration.rs:72]: these can only change on restart? (i think thats good, just wondering)

**Review by guswynn — APPROVED** [review]
looks good!

File: src/storage-client/src/controller/rehydration.rs (hunk 2 above)
Comment (pH14) [reply src/storage-client/src/controller/rehydration.rs:72]: yes -- it'll only change on restart per replica

File: src/sql/src/session/vars.rs (hunk 1 above)
Comment (pH14) [reply src/sql/src/session/vars.rs:1124]: honestly not sure. I didn't find much for http/2 keep-alive standards, and grpc defaults to not using keep-alives (with baked in defaults of interval=2h, timeout=20s if you turn it on)

**Review by antiguru — APPROVED** [review]
Had a quick look at the compute changes, and it seems to be fine.

**Comment by pH14** [comment]
Thanks for taking a look!

**Comment by benesch** [comment]
Thanks so much for tackling this one, @pH14!

**Pull request merged**